src/clj/rems/db/applications.clj

```diff
@@ -1037,6 +1044,37 @@
                               :eventdata (event->json event)})
   nil)
 
+(defn add-application-created-event! [{:keys [application-id catalogue-item-ids time actor]}]
+  (let [items (get-catalogue-items catalogue-item-ids)]
+    (assert (= 1 (count (distinct (mapv :wfid items)))) "catalogue items did not have the same workflow")
+    (assert (= 1 (count (distinct (mapv :formid items)))) "catalogue items did not have the same form")
+    (let [workflow-id (:wfid (first items))
+          form-id (:formid (first items))
+          workflow (-> (:workflow (workflow/get-workflow workflow-id))
+                       (update :type keyword))
+          licenses (get-application-licenses {:id application-id
+                                              :applicantuserid actor
+                                              :start time
+                                              :wfid workflow-id}
+                                             catalogue-item-ids)]
+      (assert (= :workflow/dynamic (:type workflow))
+              (str "workflow type was " (:type workflow))) ; TODO: support other workflows
+      (add-dynamic-event! {:event/type :application.event/created
+                           :event/time time
+                           :event/actor actor
+                           :application/id application-id
```

Comment:
Shouldn't there be a composite value under `:application` key and in it the `:id`, `:resources`, `licenses` etc.? Why is everything top-level and not hierarchical?

Reply:
Because https://github.com/CSCfi/rems/blob/master/docs/architecture/003-namespaced-keywords.md

Reply:
I guess we need to have another discussion about this because this is not exactly what I intended.